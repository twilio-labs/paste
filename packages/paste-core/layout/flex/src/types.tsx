import {ResponsiveValue} from '@twilio-paste/styling-library';
import {LayoutProps, MarginProps, PaddingProps} from '@twilio-paste/style-props';

export type DisplayOptions = 'flex' | 'inline-flex';
export type Display = ResponsiveValue<DisplayOptions>;
export type VerticalAlignOptions = 'top' | 'center' | 'bottom' | 'stretch';
export type VerticalAlign = ResponsiveValue<VerticalAlignOptions>;
export type HorizontalAlignOptions = 'left' | 'center' | 'right' | 'around' | 'between';
export type HorizontalAlign = ResponsiveValue<HorizontalAlignOptions>;
export type VerticalOptions = boolean;
export type Vertical = ResponsiveValue<VerticalOptions>;
export type GrowOptions = boolean | number;
export type Grow = ResponsiveValue<GrowOptions>;
export type ShrinkOptions = boolean | number;
export type Shrink = ResponsiveValue<ShrinkOptions>;
export type BasisOptions = string | number;
export type Basis = ResponsiveValue<BasisOptions>;
export type WrapOptions = boolean;
export type Wrap = ResponsiveValue<WrapOptions>;

export interface FlexProps extends LayoutProps, MarginProps, PaddingProps {
  as?: keyof JSX.IntrinsicElements;
  children?: React.ReactNode;
  display?: Display;
  vertical?: Vertical;
  vAlignContent?: VerticalAlign;
  hAlignContent?: HorizontalAlign;
  grow?: Grow;
  shrink?: Shrink;
  basis?: Basis;
  wrap?: Wrap;
}
