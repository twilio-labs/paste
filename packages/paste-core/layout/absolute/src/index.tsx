import * as PropTypes from 'prop-types';
import styled from '@emotion/styled';
import {top, right, bottom, left, zIndex, styleFn} from 'styled-system';
import {Box, BoxProps} from '@twilio-paste/box';

const LEFT_OPTIONS = [
  'fill',
  'left',
  'left_top',
  'left_bottom',
  'left_fill',
  'top_left',
  'top_fill',
  'bottom_left',
  'bottom_fill',
];
const RIGHT_OPTIONS = [
  'fill',
  'right',
  'right_top',
  'right_bottom',
  'right_fill',
  'top_right',
  'top_fill',
  'bottom_right',
  'bottom_fill',
];
const TOP_OPTIONS = [
  'fill',
  'top',
  'top_right',
  'top_left',
  'top_fill',
  'right_top',
  'right_fill',
  'left_top',
  'left_fill',
];
const BOTTOM_OPTIONS = [
  'fill',
  'bottom',
  'bottom_right',
  'bottom_left',
  'bottom_fill',
  'right_bottom',
  'right_fill',
  'left_bottom',
  'left_fill',
];

export type Presets =
  | 'fill'
  | 'top'
  | 'top_left'
  | 'top_right'
  | 'top_fill'
  | 'bottom'
  | 'bottom_left'
  | 'bottom_right'
  | 'bottom_fill'
  | 'left'
  | 'left_top'
  | 'left_bottom'
  | 'left_fill'
  | 'right'
  | 'right_top'
  | 'right_bottom'
  | 'right_fill';

export interface AbsoluteProps extends BoxProps {
  preset?: Presets;
}

interface PositionStyles {
  top?: 0;
  right?: 0;
  bottom?: 0;
  left?: 0;
}

const preset: styleFn = ({preset: _preset}) => {
  const positionStyles: PositionStyles = {};

  if (_preset == null) {
    return positionStyles;
  }

  if (LEFT_OPTIONS.includes(_preset)) {
    positionStyles.left = 0;
  }
  if (RIGHT_OPTIONS.includes(_preset)) {
    positionStyles.right = 0;
  }
  if (TOP_OPTIONS.includes(_preset)) {
    positionStyles.top = 0;
  }
  if (BOTTOM_OPTIONS.includes(_preset)) {
    positionStyles.bottom = 0;
  }
  return positionStyles;
};

const Absolute = styled(Box)<AbsoluteProps>(preset, zIndex, top, right, bottom, left, {
  position: 'absolute',
});

Absolute.defaultProps = {
  preset: 'fill',
};

if (process.env.NODE_ENV === 'development') {
  Absolute.propTypes = {
    preset: PropTypes.oneOf([
      'fill',
      'top',
      'top_left',
      'top_right',
      'top_fill',
      'bottom',
      'bottom_left',
      'bottom_right',
      'bottom_fill',
      'left',
      'left_top',
      'left_bottom',
      'left_fill',
      'right',
      'right_top',
      'right_bottom',
      'right_fill',
    ]),
    top: PropTypes.number,
    right: PropTypes.number,
    bottom: PropTypes.number,
    left: PropTypes.number,
  };
}

export {Absolute};
