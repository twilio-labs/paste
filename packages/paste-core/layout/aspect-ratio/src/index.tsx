import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import {styled} from '@twilio-paste/styling-library';

export interface AspectRatioProps {
  ratio: string;
  children: NonNullable<React.ReactNode>;
}

const RATIO_REGEX = /^(\d+:\d*)$/;

const isCorrectPattern = (ratio: string): boolean => RATIO_REGEX.test(ratio);

const handlePropValidation = ({ratio}: AspectRatioProps): void => {
  const hasRatio = ratio != null && ratio !== '';

  if (!hasRatio) {
    throw new Error(`[Paste: AspectRatio] Missing 'ratio' prop.`);
  }

  if (!isCorrectPattern(ratio)) {
    throw new Error(`[Paste: AspectRatio] 'ratio' is invalid. Use a colon separated number pattern (4:3).`);
  }
};

const AspectRatioContainer = styled.div`
  position: relative;

  embed,
  iframe,
  object,
  video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
  }
`;

const AspectRatio = React.forwardRef<HTMLDivElement, AspectRatioProps>((props, ref) => {
  handlePropValidation(props);

  const aspectArray = props.ratio.split(':').map(Number);
  const aspectPercent = (aspectArray[1] / aspectArray[0]) * 100;

  return (
    <AspectRatioContainer ref={ref} style={{paddingBottom: `${aspectPercent}%`}}>
      <Box position="absolute" top={0} right={0} bottom={0} left={0}>
        {props.children}
      </Box>
    </AspectRatioContainer>
  );
});

AspectRatio.displayName = 'AspectRatio';

if (process.env.NODE_ENV === 'development') {
  AspectRatio.propTypes = {
    children: PropTypes.node.isRequired,
    ratio: PropTypes.string.isRequired,
  };
}

export {AspectRatio};
