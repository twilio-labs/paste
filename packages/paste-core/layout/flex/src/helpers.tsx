import {ValueOf} from '@twilio-paste/types';
import {
  Basis,
  BasisOptions,
  FlexProps,
  GrowOptions,
  HorizontalAlignOptions,
  ShrinkOptions,
  VerticalAlignOptions,
  VerticalOptions,
  WrapOptions,
} from './types';

// Gets grow and returns as an array of grow, the value of grow, 1, or 0
export const getGrow = ({grow}: FlexProps): number | number[] => {
  if (Array.isArray(grow)) {
    return (grow as GrowOptions[]).map((value: GrowOptions) => {
      return Number(value);
    });
  }

  if (typeof grow === 'number') {
    return grow;
  }

  if (grow) {
    return 1;
  }

  return 0;
};

// Gets shrink and returns as an array of shrink, the value of shrink, 1, or 0
export const getShrink = ({shrink, basis}: FlexProps): number | number[] => {
  if (Array.isArray(shrink)) {
    return (shrink as ShrinkOptions[]).map((value: ShrinkOptions) => {
      return Number(value);
    });
  }

  if (typeof shrink === 'number') {
    return shrink;
  }

  if (typeof shrink === 'boolean') {
    return shrink ? 1 : 0;
  }

  if (basis && basis !== 'auto') {
    return 0;
  }

  return 1;
};

// Function for basis. Adds 'px' if basis is a number.
export const getSuffix = (item: Basis): string => {
  const suffix = typeof item === 'number' || String(Number.parseInt(item as string, 10)) === item ? 'px' : '';
  return item + suffix;
};

// Gets basis and returns as an array of basis, the value of basis, or auto
export const getBasis = ({basis}: FlexProps): string | string[] => {
  if (Array.isArray(basis)) {
    return (basis as BasisOptions[]).map((value: BasisOptions) => {
      return getSuffix(value);
    });
  }

  if (basis) {
    return getSuffix(basis);
  }

  return 'auto';
};

// Gets vertical and returns as an array of column/row, column, or row
type GetVerticalReturn = 'column' | 'row';
export const getVertical = ({vertical}: FlexProps): GetVerticalReturn | GetVerticalReturn[] => {
  if (Array.isArray(vertical)) {
    return (vertical as VerticalOptions[]).map((value: VerticalOptions) => {
      if (typeof value === 'boolean') {
        return value === true ? 'column' : 'row';
      }
      return value;
    });
  }

  if (vertical) {
    return 'column';
  }

  return 'row';
};

// Gets wrap and returns as an array of wrap/nowrap, wrap, or nowrap
type GetWrapReturn = 'wrap' | 'nowrap';
export const getWrap = ({wrap}: FlexProps): GetWrapReturn | GetWrapReturn[] => {
  if (Array.isArray(wrap)) {
    return (wrap as WrapOptions[]).map((value: WrapOptions) => {
      if (typeof value === 'boolean') {
        return value === true ? 'wrap' : 'nowrap';
      }
      return value;
    });
  }

  if (wrap) {
    return 'wrap';
  }

  return 'nowrap';
};

// Key value pairs for vertical alignments
const RemapedVerticalAlignments = {
  top: 'flex-start',
  center: 'center',
  bottom: 'flex-end',
  stretch: 'stretch',
} as const;

// Gets vAlignContent and returns array of verical key value pairs, a verical key value pair, or flex-start
export const vAlignToProps = ({
  vAlignContent,
}: FlexProps): ValueOf<typeof RemapedVerticalAlignments> | ValueOf<typeof RemapedVerticalAlignments>[] => {
  if (Array.isArray(vAlignContent)) {
    return (vAlignContent as VerticalAlignOptions[]).map(value => RemapedVerticalAlignments[value]);
  }

  if (vAlignContent) {
    return RemapedVerticalAlignments[vAlignContent as VerticalAlignOptions];
  }

  return 'flex-start';
};

// Key value pairs for horizontal alignments
const RemapedHorizontalAlignments = {
  left: 'flex-start',
  center: 'center',
  right: 'flex-end',
  around: 'space-around',
  between: 'space-between',
} as const;

// Gets hAlignContent and returns array of horizontal key value pairs, a horizontal key value pair, or flex-start
// ideally this would be typed as ValueOf<typeof RemapedHorizontalAlignments> | ValueOf<typeof RemapedHorizontalAlignments>[]
// but space-between isn't a valid alignItems value and I don't know how to conditionally type things.
export const hAlignToProps = ({hAlignContent}: FlexProps): any => {
  if (Array.isArray(hAlignContent)) {
    return (hAlignContent as HorizontalAlignOptions[]).map(value => RemapedHorizontalAlignments[value]);
  }

  if (hAlignContent) {
    return RemapedHorizontalAlignments[hAlignContent as HorizontalAlignOptions];
  }

  return 'flex-start';
};
