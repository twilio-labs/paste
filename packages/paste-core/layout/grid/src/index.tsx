export {Grid} from './Grid';
export {Column} from './Column';
export {GridProps, ColumnProps} from './types';
