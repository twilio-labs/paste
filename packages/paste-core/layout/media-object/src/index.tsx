import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, BoxProps} from '@twilio-paste/box';
import {Space, isMarginTokenProp} from '@twilio-paste/style-props';

export interface MediaObjectProps extends Pick<BoxProps, 'as' | 'marginTop' | 'marginBottom'> {
  children: NonNullable<React.ReactNode>;
  verticalAlign?: 'center' | 'top';
}

const MediaObject = React.forwardRef<HTMLElement, MediaObjectProps>(
  ({as = 'span', children, marginBottom, marginTop, verticalAlign = 'top'}, ref) => {
    return (
      <Box
        as={as}
        display="flex"
        alignItems={verticalAlign === 'top' ? 'flex-start' : 'center'}
        marginBottom={marginBottom}
        marginTop={marginTop}
        ref={ref}
      >
        {children}
      </Box>
    );
  }
);

MediaObject.displayName = 'MediaObject';
if (process.env.NODE_ENV === 'development') {
  MediaObject.propTypes = {
    children: PropTypes.node.isRequired,
    verticalAlign: PropTypes.oneOf(['center', 'top']),
  };
}

export interface MediaFigureProps extends Pick<BoxProps, 'as'> {
  align?: 'start' | 'end';
  children: NonNullable<React.ReactNode>;
  spacing?: Space;
}

const MediaFigure = React.forwardRef<HTMLElement, MediaFigureProps>(({as = 'span', children, align, spacing}, ref) => {
  return (
    <Box
      as={as}
      display="flex"
      flexShrink={0}
      marginLeft={align === 'end' ? spacing : undefined}
      marginRight={align === 'start' ? spacing : undefined}
      ref={ref}
    >
      {children}
    </Box>
  );
});

MediaFigure.displayName = 'MediaFigure';
MediaFigure.defaultProps = {
  align: 'start',
  spacing: 'space0',
};
if (process.env.NODE_ENV === 'development') {
  MediaFigure.propTypes = {
    align: PropTypes.oneOf(['start', 'end']),
    children: PropTypes.node.isRequired,
    spacing: isMarginTokenProp,
  };
}

// export type MediaBodyProps = Pick<BoxProps, 'as'>;
export interface MediaBodyProps extends Pick<BoxProps, 'as'> {
  children: NonNullable<React.ReactNode>;
}

const MediaBody = React.forwardRef<HTMLElement, MediaBodyProps>(({as = 'span', children}, ref) => {
  return (
    <Box as={as} flex={1} minWidth="size0" ref={ref}>
      {children}
    </Box>
  );
});

MediaBody.displayName = 'MediaBody';
if (process.env.NODE_ENV === 'development') {
  MediaBody.propTypes = {
    children: PropTypes.node.isRequired,
  };
}

export {MediaObject, MediaFigure, MediaBody};
