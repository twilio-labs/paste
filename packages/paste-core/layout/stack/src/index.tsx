import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUIDSeed} from '@twilio-paste/uid-library';
import {ResponsiveValue} from '@twilio-paste/styling-library';
import {
  isSpaceTokenProp,
  ResponsiveProp,
  LayoutProps,
  FlexboxProps,
  MarginProps,
  SpaceOptions,
} from '@twilio-paste/style-props';
import {Box, BoxElementProps, safelySpreadBoxProps} from '@twilio-paste/box';

type StackChildMargins = Pick<MarginProps, 'marginRight' | 'marginBottom'>;
type DisplayOptions = 'block' | 'flex';
type DisplayValue = ResponsiveValue<DisplayOptions>;
type StackOrientationOptions = 'horizontal' | 'vertical';
export type StackOrientation = ResponsiveValue<StackOrientationOptions>;

interface StackStyleProps extends Pick<LayoutProps, 'display'>, Pick<FlexboxProps, 'alignItems' | 'flexWrap'> {}
export interface StackProps extends BoxElementProps {
  orientation: StackOrientation;
  spacing: SpaceOptions;
}

export const getStackDisplay = (orientation: StackOrientation): DisplayValue => {
  if (Array.isArray(orientation)) {
    return orientation.map(value => {
      if (value === 'horizontal') {
        return 'flex';
      }

      return 'block';
    });
  }

  if (orientation === 'horizontal') {
    return 'flex';
  }

  return 'block';
};

export const getStackStyles = (orientation: StackOrientation): StackStyleProps => {
  const styles: StackStyleProps = {
    display: getStackDisplay(orientation),
    alignItems: 'center',
    flexWrap: 'wrap',
  };

  return styles;
};

export const getStackChildMargins = (orientation: StackOrientation, spacing: SpaceOptions): StackChildMargins => {
  let styles = {};

  if (Array.isArray(orientation)) {
    const marginRight: SpaceOptions[] = [];
    const marginBottom: SpaceOptions[] = [];

    orientation.forEach((value, i) => {
      marginRight[i] = value === 'horizontal' ? spacing : 'space0';
      marginBottom[i] = value === 'horizontal' ? 'space0' : spacing;
    });

    styles = {
      marginRight,
      marginBottom,
    };
  }

  if (orientation === 'horizontal') {
    styles = {marginRight: spacing};
  }

  if (orientation === 'vertical') {
    styles = {marginBottom: spacing};
  }

  return styles;
};

const Stack = React.forwardRef<HTMLDivElement, StackProps>(({children, orientation, spacing, ...props}, ref) => {
  const [childrenCount, validChildren] = React.useMemo(
    () => [
      React.Children.count(children),
      React.Children.toArray(children).filter(child => React.isValidElement(child) || typeof child === 'string'),
    ],
    [children]
  );
  const stackStyles = React.useMemo(() => getStackStyles(orientation), [orientation]);
  const childMargins = React.useMemo(() => getStackChildMargins(orientation, spacing), [orientation, spacing]);
  const keySeed = useUIDSeed();

  return (
    <Box {...safelySpreadBoxProps(props)} {...stackStyles} ref={ref}>
      {validChildren.map((child, index) => {
        return (
          <Box {...(childrenCount !== index + 1 ? childMargins : null)} key={keySeed(`stack-${index}`)}>
            {child}
          </Box>
        );
      })}
    </Box>
  );
});

Stack.displayName = 'Stack';

if (process.env.NODE_ENV === 'development') {
  Stack.propTypes = {
    orientation: ResponsiveProp(PropTypes.oneOf(['horizontal', 'vertical'])).isRequired,
    spacing: isSpaceTokenProp,
  };
}

export {Stack};
