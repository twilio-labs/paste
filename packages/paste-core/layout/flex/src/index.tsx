import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {FlexboxProps, ResponsiveProp} from '@twilio-paste/style-props';
import {FlexProps} from './types';
import {getGrow, getShrink, getBasis, getVertical, getWrap, hAlignToProps, vAlignToProps} from './helpers';

const getFlexStyles = (props: FlexProps): FlexboxProps => {
  const styles: FlexboxProps = {
    justifyContent: props.vertical ? vAlignToProps(props) : hAlignToProps(props),
    alignItems: props.vertical ? hAlignToProps(props) : vAlignToProps(props),
  };

  if (props.grow || props.shrink || props.basis) {
    styles.flexGrow = getGrow(props);
    styles.flexShrink = getShrink(props);
    styles.flexBasis = getBasis(props);
  }

  if (props.vertical) {
    styles.flexDirection = getVertical(props);
  }

  if (props.wrap) {
    styles.flexWrap = getWrap(props);
  }

  return styles;
};

const Flex: React.FC<FlexProps> = ({
  as,
  basis,
  children,
  display,
  hAlignContent,
  grow,
  marginTop,
  marginRight,
  marginBottom,
  marginLeft,
  paddingTop,
  paddingRight,
  paddingBottom,
  paddingLeft,
  maxWidth,
  minWidth = 'size0',
  shrink,
  vertical,
  vAlignContent,
  width,
  wrap,
  ...props
}) => {
  const FlexStyles = React.useMemo(
    () => getFlexStyles({basis, hAlignContent, grow, shrink, vertical, vAlignContent, wrap}),
    [basis, hAlignContent, grow, shrink, vertical, vAlignContent, wrap]
  );
  return (
    <Box
      {...FlexStyles}
      {...safelySpreadBoxProps(props)}
      as={as}
      display={display}
      marginTop={marginTop}
      marginRight={marginRight}
      marginBottom={marginBottom}
      marginLeft={marginLeft}
      paddingTop={paddingTop}
      paddingRight={paddingRight}
      paddingBottom={paddingBottom}
      paddingLeft={paddingLeft}
      maxWidth={maxWidth}
      minWidth={minWidth}
      width={width}
    >
      {children}
    </Box>
  );
};

Flex.displayName = 'Flex';
Flex.defaultProps = {
  display: 'flex',
};

if (process.env.NODE_ENV === 'development') {
  Flex.propTypes = {
    as: PropTypes.string as any,
    display: ResponsiveProp(PropTypes.oneOf(['flex', 'inline-flex'])),
    vertical: ResponsiveProp(PropTypes.bool),
    vAlignContent: ResponsiveProp(PropTypes.oneOf(['top', 'center', 'bottom', 'stretch'])),
    hAlignContent: ResponsiveProp(PropTypes.oneOf(['left', 'center', 'right', 'around', 'between'])),
    grow: ResponsiveProp(PropTypes.oneOfType([PropTypes.bool, PropTypes.number])),
    shrink: ResponsiveProp(PropTypes.oneOfType([PropTypes.bool, PropTypes.number])),
    basis: ResponsiveProp(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
    wrap: ResponsiveProp(PropTypes.bool),
  };
}

export {Flex};
