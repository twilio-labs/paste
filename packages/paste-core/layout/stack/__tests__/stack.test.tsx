import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {Card} from '@twilio-paste/card';
import {Theme} from '@twilio-paste/theme';
import {getStackDisplay, getStackStyles, getStackChildMargins, Stack, StackOrientation} from '../src';

describe('Stack Unit Tests', () => {
  const mockHorizontalOrientation = 'horizontal';
  const mockVerticalOrientation = 'vertical';
  const mockResponsiveOrientation: StackOrientation = ['vertical', 'horizontal', 'vertical'];
  const mockSpace = 'space40';
  const horizontalStyles = {display: 'flex', alignItems: 'center', flexWrap: 'wrap'};
  const verticalStyles = {display: 'block', alignItems: 'center', flexWrap: 'wrap'};
  const marginRightStyles = {marginRight: 'space40'};
  const marginBottomStyles = {marginBottom: 'space40'};
  const marginReponsiveStyles = {
    marginBottom: ['space40', 'space0', 'space40'],
    marginRight: ['space0', 'space40', 'space0'],
  };

  it('should return display: flex', (): void => {
    expect(getStackDisplay(mockHorizontalOrientation)).toStrictEqual('flex');
  });

  it('should return display: block', (): void => {
    expect(getStackDisplay(mockVerticalOrientation)).toStrictEqual('block');
  });

  it('should return a reponsive display', (): void => {
    expect(getStackDisplay(mockResponsiveOrientation)).toStrictEqual(['block', 'flex', 'block']);
  });

  it('should return horizontal styles', (): void => {
    expect(getStackStyles(mockHorizontalOrientation)).toStrictEqual(horizontalStyles);
  });

  it('should return vertical styles', (): void => {
    expect(getStackStyles(mockVerticalOrientation)).toStrictEqual(verticalStyles);
  });

  it('should return marginRight styles', (): void => {
    expect(getStackChildMargins(mockHorizontalOrientation, mockSpace)).toStrictEqual(marginRightStyles);
  });

  it('should return marginBottom styles', (): void => {
    expect(getStackChildMargins(mockVerticalOrientation, mockSpace)).toStrictEqual(marginBottomStyles);
  });

  it('should return responsive margin styles', (): void => {
    expect(getStackChildMargins(mockResponsiveOrientation, mockSpace)).toStrictEqual(marginReponsiveStyles);
  });
});

const MockVerticalStack: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <Stack orientation="vertical" spacing="space60">
        <Card>Card one</Card>
        <Card>Card two</Card>
      </Stack>
    </Theme.Provider>
  );
};

const MockHorizontalStack: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <Stack orientation="horizontal" spacing="space60">
        <Card>Card one</Card>
        <Card>Card two</Card>
      </Stack>
    </Theme.Provider>
  );
};

const MockResponsiveStack: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <Stack orientation={['horizontal', 'vertical', 'horizontal']} spacing="space60">
        <Card>Card one</Card>
        <Card>Card two</Card>
      </Stack>
    </Theme.Provider>
  );
};

const MockHeaderStack: React.FC = () => {
  return (
    <Theme.Provider theme="console">
      <Stack
        as="header"
        orientation="vertical"
        spacing="space60"
        data-testid="header"
        id="foo"
        title="foo"
        className="foo"
      >
        <Card>Card one</Card>
        <Card>Card two</Card>
      </Stack>
    </Theme.Provider>
  );
};

describe('Stack', () => {
  it('should render a vertical stack', () => {
    const {asFragment} = render(<MockVerticalStack />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a horizontal stack', () => {
    const {asFragment} = render(<MockHorizontalStack />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a responsive stack', () => {
    const {asFragment} = render(<MockResponsiveStack />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render as a header', () => {
    render(<MockHeaderStack />);
    expect(screen.getByTestId('header').tagName).toEqual('HEADER');
  });

  it('should render with an id', () => {
    render(<MockHeaderStack />);
    expect(screen.getByTestId('header').getAttribute('id')).toEqual('foo');
  });

  it('should render with a title', () => {
    render(<MockHeaderStack />);
    expect(screen.getByTestId('header').getAttribute('title')).toEqual('foo');
  });

  it('should render without a className', () => {
    render(<MockHeaderStack />);
    expect(screen.getByTestId('header').getAttribute('class')).not.toEqual('foo');
  });
});
