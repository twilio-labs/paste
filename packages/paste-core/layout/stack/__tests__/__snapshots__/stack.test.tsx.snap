// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack should render a horizontal stack 1`] = `
<DocumentFragment>
  .emotion-5 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: rgb(228,231,233);
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: rgb(255,255,255);
}

.emotion-3 {
  box-sizing: border-box;
}

.emotion-4 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-1 {
  box-sizing: border-box;
  margin-right: 1.25rem;
}

<div
    class="emotion-5"
  >
    <div
      class="emotion-4"
      display="flex"
    >
      <div
        class="emotion-1"
      >
        <article
          class="emotion-0"
        >
          Card one
        </article>
      </div>
      <div
        class="emotion-3"
      >
        <article
          class="emotion-0"
        >
          Card two
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Stack should render a responsive stack 1`] = `
<DocumentFragment>
  .emotion-5 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: rgb(228,231,233);
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: rgb(255,255,255);
}

.emotion-3 {
  box-sizing: border-box;
}

.emotion-4 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media screen and (min-width:25rem) {
  .emotion-4 {
    display: block;
  }
}

@media screen and (min-width:64rem) {
  .emotion-4 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

.emotion-1 {
  box-sizing: border-box;
  margin-right: 1.25rem;
  margin-bottom: 0;
}

@media screen and (min-width:25rem) {
  .emotion-1 {
    margin-right: 0;
    margin-bottom: 1.25rem;
  }
}

@media screen and (min-width:64rem) {
  .emotion-1 {
    margin-right: 1.25rem;
    margin-bottom: 0;
  }
}

<div
    class="emotion-5"
  >
    <div
      class="emotion-4"
      display="flex,block,flex"
    >
      <div
        class="emotion-1"
      >
        <article
          class="emotion-0"
        >
          Card one
        </article>
      </div>
      <div
        class="emotion-3"
      >
        <article
          class="emotion-0"
        >
          Card two
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Stack should render a vertical stack 1`] = `
<DocumentFragment>
  .emotion-5 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
  font-variant-numeric: tabular-nums;
}

.emotion-4 {
  box-sizing: border-box;
  display: block;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-1 {
  box-sizing: border-box;
  margin-bottom: 1.25rem;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: rgb(228,231,233);
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: rgb(255,255,255);
}

.emotion-3 {
  box-sizing: border-box;
}

<div
    class="emotion-5"
  >
    <div
      class="emotion-4"
      display="block"
    >
      <div
        class="emotion-1"
      >
        <article
          class="emotion-0"
        >
          Card one
        </article>
      </div>
      <div
        class="emotion-3"
      >
        <article
          class="emotion-0"
        >
          Card two
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
